### PR TITLE
support named routes

### PR DESCRIPTION
**Pull request by saeideng**
alternative to #100

I need more tests and feedback 
don't merge

### CODE REVIEW

**Comment by dereuromark** [comment]
@lorro Please test this one.

This looks quite good already. :+1:

**Comment by dereuromark** [comment]
Would be nice if someone could also add some basic docs on how to use this new feature.

**Comment by saeideng** [comment]
just remains throwing a exception when a named route does not exist 
i think cake itself should throws exception for this 
```php
Router:Connect('..../{title}',['controller'=>..., 'action'=>...., ],['_name'=>..., .....]);
Router::url(['_name'=>....])
````
you can see no `title` value in input , so , cake throws a exception and is acceptable 
but for this helper ( `hasAccess` ) should we behave strict ?

I'll test for find a good solution but any feedback appreciated

**Comment by dereuromark** [comment]
I would say that as a first step we dont have to be too strict here maybe?

**Comment by saeideng** [comment]
>I would say that as a first step we dont have to be too strict here maybe?

I prefer strict mode , I'll add another commit for this purpose if you agree

File: tests/TestCase/View/Helper/AuthUserHelperTest.php [hunk 1]
```diff
@@ -95,6 +128,67 @@ public function testIsAuthorizedNotLoggedIn() {
...
+			'/edit/*',
+			['action' => 'edit'],
+			['_name' => 'Tags::edit']
+		);
+
+		$request = ['_name' => 'InvalidName'];
+
+		$this->expectException(\Cake\Routing\Exception\MissingRouteException::class);
+
+		$result = $this->AuthUserHelper->hasAccess($request);
```
Comment (dereuromark) [inline tests/TestCase/View/Helper/AuthUserHelperTest.php]: $result is unused here

File: tests/TestCase/View/Helper/AuthUserHelperTest.php [hunk 2]
```diff
@@ -95,6 +128,67 @@ public function testIsAuthorizedNotLoggedIn() {
...
+
+		Router::connect(
+			'/edit/*',
+			['action' => 'edit'],
+			['_name' => 'Tags::edit']
+		);
+
+		$request = ['_name' => 'InvalidName'];
+
+		$this->expectException(\Cake\Routing\Exception\MissingRouteException::class);
```
Comment (dereuromark) [inline tests/TestCase/View/Helper/AuthUserHelperTest.php]: use statement instead maybe?

**Comment by dereuromark** [comment]
Can we move forward here?

File: src/View/Helper/AuthUserHelper.php [hunk 3]
```diff
@@ -42,13 +44,31 @@ public function __construct(View $View, array $config = []) {
...
+				if (!isset($defaults['action']) || !isset($defaults['controller'])) {
+					throw new Exception('Controller or action name could not be null.');
+				}
+				$url = [
+					'prefix' => !empty($defaults['prefix']) ? $defaults['prefix'] : null,
+					'plugin' => !empty($defaults['plugin']) ? $defaults['plugin'] : null,
+					'controller' => $defaults['controller'],
+					'action' => $defaults['action'],
+				];
+            } else {
```
Comment (dereuromark) [inline src/View/Helper/AuthUserHelper.php]: You could invert this and have a cleaner flow.

**Comment by dereuromark** [comment]
@lorro Is this as you want it?

**Comment by saeideng** [comment]
@dereuromark  this merged before completion , I'll Open another PR very soon 
I think we have to use strict mode for now

**Comment by dereuromark** [comment]
I wont release a tag here yet (minor), but I wanted to start documentation on it.

**Comment by lorro** [comment]
Sorry for my unresponsiveness... Been away for a while.
I'll try this out asap.

**Pull request merged**